File: exercism/haskell/rotational-cipher/src/RotationalCipher.hs
module RotationalCipher
  ( rotate
  )
where

import Data.Char

rotate :: Int -> String -> String
rotate offset = fmap translate
  where
    changeByBase chBase ch =
      chr $ ((ord ch - baseOrd + offset) `rem` 26) + baseOrd
      where
        baseOrd = ord chBase
    translate ch
      | isAsciiUpper ch = changeByBase 'A' ch
      | isAsciiLower ch = changeByBase 'a' ch
      | otherwise = ch

File: exercism/kotlin/rotational-cipher/src/main/kotlin/RotationalCipher.kt
class RotationalCipher(val offset: Int) {
    fun overInt(f: (Int) -> Int): (Char) -> Char = { ch ->
        when {
            ch.isLowerCase() -> ('a' + f(ch - 'a'))
            ch.isUpperCase() -> ('A' + f(ch - 'A'))
            else -> ch
        }
    }

    fun encode(text: String): String =
        text.map(overInt { (it + offset).rem(26) }).joinToString("")
}
